src/librustdoc/html/render/span_map.rs

```diff
@@ -0,0 +1,209 @@
+use crate::clean;
+use crate::html::sources;
+
+use rustc_data_structures::fx::FxHashMap;
+use rustc_hir::def::{DefKind, Res};
+use rustc_hir::def_id::DefId;
+use rustc_hir::intravisit::{self, NestedVisitorMap, Visitor};
+use rustc_hir::{ExprKind, GenericParam, GenericParamKind, HirId, Mod, Node};
+use rustc_middle::ty::TyCtxt;
+use rustc_span::Span;
+
+use std::path::{Path, PathBuf};
+
+/// This enum allows us to store two different kinds of information:
+///
+/// In case the `span` definition comes from the same crate, we can simply get the `span` and use
+/// it as is.
+///
+/// Otherwise, we store the definition `DefId` and will generate a link to the documentation page
+/// instead of the source code directly.
+#[derive(Debug)]
+crate enum LinkFromSrc {
+    Local(clean::Span),
+    External(DefId),
+}
+
+/// This struct is used only as index in the `span_map`, not as [`Span`]! `Span`s contain
+/// some extra information (the syntax context) we don't need. **Do not convert this type back to
+/// `Span`!!!**
+#[derive(Clone, Copy, Hash, PartialEq, Eq, Debug)]
+crate struct LightSpan {
+    crate lo: u32,
+    crate hi: u32,
+}
+
+impl LightSpan {
+    /// Before explaining what this method does, some global explanations on rust's `Span`:
+    ///
+    /// Each source code file is stored in the source map in the compiler and has a
+    /// `lo` and a `hi` (lowest and highest bytes in this source map which can be seen as one huge
+    /// string to simplify things). So in this case, this represents the starting byte of the
+    /// current file. It'll be used later on to retrieve the "definition span" from the
+    /// `span_correspondance_map` (which is inside `context`).
+    ///
+    /// This when we transform the "span" we have from reading the input into a "span" which can be
+    /// used as index to the `span_correspondance_map` to get the definition of this item.
+    ///
+    /// So in here, `file_span_lo` is representing the "lo" byte in the global source map, and to
+    /// make our "span" works in there, we simply add `file_span_lo` to our values.
+    crate fn new_in_file(file_span_lo: u32, lo: u32, hi: u32) -> Self {
```

Comment:
This seems awfully complicated for a concept that already exists in the compiler. Why not reuse Span?

Reply:
Because we don't keep the `SyntaxContext` in the "span map" (makes the keys smaller), but otherwise it's basically the same.

Reply:
It could be simplified a bit if we put the creation of the `LightSpan` in a method on `Classifier` if you prefer? The first argument would then be provided by the `Classifier` directly. As for `lo` and `hi`, they are the position in the source file string.

Reply:
I created a `Classifier::new_light_span` method which makes more convenient to use (no need to pass `file_span_lo` every time).

Reply:
> The goal was to not perform any span operation which are checking inside the global span context, etc.

But why is that a goal? You're introducing a ton more boilerplate (including some subtle things around file starts) that already exist in Span.

Reply:
Maybe I was doing premature optimization... Originally, I didn't add `LightSpan`, I added it afterwards to improve code readability. spans are basically indexes over all source codes of a crate, so even by using `Span`, I'd still need to save the starting point of the file and the `SyntaxContext` to then create a `Span` (instead of `LightSpan`) for the URL generation.

We can compare this logic with how we generate `Span` for errors in intra-doc links.

Well anyway, I can remove the whole `LightSpan` struct and use `Span` instead if you prefer?